cmd/orch-tester/orch_tester.go

```diff
@@ -549,3 +599,209 @@ func (s *streamerClient) getOrchestrators() ([]*orch, error) {
+	}
+	renditionPl, err := s.downloadMediaPlaylist(mpl.Variants[rand.Int()%len(mpl.Variants)].URI)
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	if len(renditionPl.Segments) == 0 {
+		return nil, nil, "", fmt.Errorf("no transcoded segments")
+	}
+
+	var rendSegs []*m3u8.MediaSegment
+	var sourceSegs []*m3u8.MediaSegment
+	for _, v := range renditionPl.Segments {
+		if v != nil {
+			rendSegs = append(rendSegs, v)
+		}
+	}
+	for _, v := range sourcePl.Segments {
+		if v != nil {
+			sourceSegs = append(sourceSegs, v)
+		}
+	}
+	if len(rendSegs) == 0 || len(sourceSegs) == 0 {
+		return nil, nil, "", fmt.Errorf("no segments found")
+	}
+
+	seqNo := r % int(math.Min(float64(len(rendSegs)), float64(len(sourceSegs))))
+	rendition, err = s.downloadSegment(fmt.Sprintf("%v/%v", s.broadcaster, rendSegs[seqNo].URI))
+	if err != nil {
+		return nil, nil, "", err
+	}
+	source, err = s.downloadSegment(fmt.Sprintf("%v/%v", s.broadcaster, sourceSegs[seqNo].URI))
+	if err != nil {
+		return nil, nil, "", err
+	}
+	return source, rendition, rendSegs[seqNo].URI, err
+}
+
+func (s *streamerClient) downloadSegment(url string) ([]byte, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := s.client.Do(req)
+	if err != nil {
+		glog.Error(err)
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		glog.Error(err)
+		return nil, err
+	}
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, fmt.Errorf("%v", string(body))
+	}
+	return body, nil
+}
+
+func (s *streamerClient) downloadMasterPlaylist(manifestID string) (*m3u8.MasterPlaylist, error) {
+	url := fmt.Sprintf("%v/stream/%v_0_0.m3u8", s.broadcaster, manifestID)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, errors.New(string(body))
+	}
+
+	mpl := m3u8.NewMasterPlaylist()
+	if err := mpl.Decode(*bytes.NewBuffer(body), true); err != nil {
+		return nil, err
+	}
+
+	return mpl, nil
+}
+
+func (s *streamerClient) downloadMediaPlaylist(uri string) (*m3u8.MediaPlaylist, error) {
+	url := fmt.Sprintf("%v/stream/%v", s.broadcaster, uri)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, errors.New(string(body))
+	}
+
+	pl, err := m3u8.NewMediaPlaylist(1, 1)
+	if err != nil {
+		return nil, err
+	}
+	if err := pl.Decode(*bytes.NewBuffer(body), true); err != nil {
+		return nil, err
+	}
+
+	for _, s := range pl.Segments {
```

Comment:
When would there be a nil MediaSegment?

Reply:
Bump on this question. I see your other [comment](https://github.com/livepeer/stream-tester/pull/68#discussion_r546906111) about filtering out nil values, but I'd like to understand what would cause nil values in the middle of the slice.

Reply:
Right, I'm not sure if there would be any in the middle of the slice but I have encoutered a few nil values at the end of the slice. So if we're positive this doesn't occur in the middle of the slice either I guess we can break once we encounter a nil value.

Reply:
I see. If it is possible for nil values to be in the middle of the slice then keeping the nil checks makes sense.